enrollment-server-onboarding-provider-innovatrics/src/main/java/com/wultra/app/onboardingserver/provider/innovatrics/InnovatricsDocumentVerificationProvider.java

```diff
@@ -55,31 +121,301 @@ public boolean shouldStoreSelfie() {
+
+        final DocumentVerificationResult result = new DocumentVerificationResult();
+        result.setUploadId(customerId);
+        result.setVerificationResult(serializeToString(response));
+        if (!rejectionReasons.isEmpty()) {
+            result.setRejectReason(serializeToString(rejectionReasons));
+        }
+        return result;
+    }
+
+    /**
+     * Parse VisualZoneInspection of a document provided by Innovatrics.
+     * @param visualZoneInspection inspection of a document by Innovatrics.
+     * @return List of reasons to reject the document.
+     */
+    private static List<String> parseVisualZoneInspection(final VisualZoneInspection visualZoneInspection) {
+        final List<String> rejectionReasons = new ArrayList<>();
+        if (visualZoneInspection == null) {
+            return rejectionReasons;
+        }
+
+        // Contains fields with a ocr confidence lower than ocr-text-field-threshold settings.
+        final List<String> lowOcrConfidenceAttributes = visualZoneInspection.getOcrConfidence().getLowOcrConfidenceTexts();
+        if (!CollectionUtils.isEmpty(lowOcrConfidenceAttributes)) {
+            rejectionReasons.add("Low OCR confidence of attributes: %s".formatted(lowOcrConfidenceAttributes));
+        }
+
+        final TextConsistency textConsistency = visualZoneInspection.getTextConsistency();
+        if (textConsistency == null) {
+            return rejectionReasons;
+        }
+
+        final TextConsistentWith textConsistentWith = textConsistency.getConsistencyWith();
+        if (textConsistentWith == null) {
+            return rejectionReasons;
+        }
+
+        final MrzConsistency mrzConsistency = textConsistentWith.getMrz();
+        if (mrzConsistency != null) {
+            final List<String> inconsistentAttributes = getCrucial(mrzConsistency.getInconsistentTexts());
+            if (!inconsistentAttributes.isEmpty()) {
+                rejectionReasons.add("Inconsistent crucial attributes with MRZ: %s".formatted(inconsistentAttributes));
+            }
+        }
+
+        final BarcodesConsistency barcodesConsistency = textConsistentWith.getBarcodes();
+        if (barcodesConsistency != null) {
+            final List<String> inconsistentAttributes = getCrucial(barcodesConsistency.getInconsistentTexts());
+            if (!inconsistentAttributes.isEmpty()) {
+                rejectionReasons.add("Inconsistent crucial attributes with barcode: %s".formatted(inconsistentAttributes));
+            }
+        }
+
+        return rejectionReasons;
+    }
+
+    /**
+     * Intersects list of attributes with CRUCIAL_ATTRIBUTES.
+     * @param attributes Attributes to do the intersection on.
+     * @return Attributes intersection.
+     */
+    private static List<String> getCrucial(final List<String> attributes) {
+        if (attributes == null) {
+            return new ArrayList<>();
+        }
+        return attributes.stream().filter(CRUCIAL_ATTRIBUTES::contains).toList();
```

Comment:
This creates an immutable list, which is good, but the line above returns mutable ones.